src/main/cpp/apple_fsnotifier.cpp

```diff
@@ -1,116 +1,145 @@
-/*
- * Initial version copied from https://github.com/JetBrains/intellij-community/blob/master/native/fsNotifier/mac/fsnotifier.c
- */
-// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
-
-// Important notes on FSEvents: https://developer.apple.com/library/archive/documentation/Darwin/Conceptual/FSEvents_ProgGuide/UsingtheFSEventsFramework/UsingtheFSEventsFramework.html
-
-/*
- * Apple specific functions.
- */
 #if defined(__APPLE__)
 
 #include "net_rubygrapefruit_platform_internal_jni_OsxFileEventFunctions.h"
 #include "generic.h"
 #include <CoreServices/CoreServices.h>
 #include <pthread.h>
-#include <strings.h>
 
-bool invalidStateDetected = false;
+class Server;
+
+static void handleEventsCallback(
+    ConstFSEventStreamRef streamRef,
+    void *clientCallBackInfo,
+    size_t numEvents,
+    void *eventPaths,
+    const FSEventStreamEventFlags eventFlags[],
+    const FSEventStreamEventId eventIds[]);
+
+static void *EventProcessingThread(void *data);
+
+class Server {
+public:
+    Server(JavaVM *jvm, JNIEnv *env, jobject watcherCallback, CFMutableArrayRef rootsToWatch, long latencyInMillis);
+    ~Server();
+
+private:
+    void run();
+    friend void *EventProcessingThread(void *data);
+
+    void handleEvents(
+        size_t numEvents,
+        char **eventPaths,
+        const FSEventStreamEventFlags eventFlags[],
+        const FSEventStreamEventId eventIds[]);
+    friend void handleEventsCallback(
+        ConstFSEventStreamRef streamRef,
+        void *clientCallBackInfo,
+        size_t numEvents,
+        void *eventPaths,
+        const FSEventStreamEventFlags eventFlags[],
+        const FSEventStreamEventId eventIds[]);
+
+    void handleEvent(JNIEnv *env, char* path, FSEventStreamEventFlags flags);
+
+    // TODO: Move this to somewhere else
+    JNIEnv* getThreadEnv();
```

Comment:
Why doesn't server keep a reference to the `JNIEnv` associated with the thread?

Reply:
It could, though then we'd need to duplicate the error handling in case the `env` is not set.

Reply:
Currently, the thread crashes since `getThreadEnv()` will return `NULL`. I wouldn't call that error handling...

Reply:
Yes, but we'll eventually have error handling implemented inside `getThreadEnv()`—which would then need to be duplicated. I don't want to do that.

Besides, even now before the thread crashes we log the fact that we couldn't get the `env` from the thread to `stderr`.

The `env` doesn't belong to the server as such, instead it belongs to the thread, and it is stored in a thread local there already. I'd argue that getting it from the thread local is the best way we have to do this.